File: day3_part1_2019.exs

defmodule CrossedWires do
  def call do
    [wire1, wire2] = File.read!("input.txt") |> String.split("\n")
    {path1, path2} = {trace_path(wire1), trace_path(wire2)}
    intersections = MapSet.intersection(MapSet.new(path1), MapSet.new(path2))
    Enum.map(intersections, &manhattan_distance(&1)) |> Enum.min()
  end

  defp trace_path(wire) do
    wire
    |> String.split(",")
    |> Enum.reduce({{0, 0}, []}, fn instruction, {{x, y}, acc} ->
      {direction, steps} = parse_instruction(instruction)
      {new_coords, path} = apply_moves({x, y}, direction, String.to_integer(steps))
      {new_coords, acc ++ path}
    end)
    |> elem(1)
  end

  defp parse_instruction(instruction) do
    {String.slice(instruction, 0, 1), String.slice(instruction, 1..-1)}
  end

  defp apply_moves(coords, direction, steps) do
    Enum.reduce(1..steps, {coords, []}, fn _, {{x, y}, acc} ->
      new_coords = case direction do
        "U" -> {x, y + 1}
        "D" -> {x, y - 1}
        "L" -> {x - 1, y}
        "R" -> {x + 1, y}
      end
      {new_coords, acc ++ [new_coords]}
    end)
  end

  defp manhattan_distance({x, y}), do: abs(x) + abs(y)
end

File: training_data/reflection-tuning/day3_part1_2019.clj
(require '[clojure.string :as str]
         '[clojure.set :as set])

(defn parse-path [path]
  (map #(vector (first %) (Integer/parseInt (subs % 1))) (str/split path #",")))

(defn generate-coords [path]
  (loop [steps path
         x 0
         y 0
         coords #{}]
    (if (empty? steps)
      coords
      (let [[dir dist] (first steps)
            new-coords (case dir
                         \R (set (map #(vector (+ x %) y) (range 1 (inc dist))))
                         \L (set (map #(vector (- x %) y) (range 1 (inc dist))))
                         \U (set (map #(vector x (+ y %)) (range 1 (inc dist))))
                         \D (set (map #(vector x (- y %)) (range 1 (inc dist)))))]
        (recur (rest steps)
               (case dir \R (+ x dist) \L (- x dist) x)
               (case dir \U (+ y dist) \D (- y dist) y)
               (set/union coords new-coords))))))

(defn manhattan-distance [[x y]]
  (+ (Math/abs x) (Math/abs y)))

(defn solve-wire-crossing []
  (let [[wire1 wire2] (str/split-lines (slurp "input.txt"))
        path1 (parse-path wire1)
        path2 (parse-path wire2)
        coords1 (generate-coords path1)
        coords2 (generate-coords path2)
        intersections (set/intersection coords1 coords2)]
    (apply min (map manhattan-distance intersections))))

(println (solve-wire-crossing))
